File: ALittle/Module/ALittleIDE/Other/GameLibrary/Std/Script/Std/Loop/LoopGroup.lua

module("ALittle", package.seeall)

local ___rawset = rawset
local ___pairs = pairs
local ___ipairs = ipairs


assert(ALittle.LoopObject, " extends class:ALittle.LoopObject is nil")
LoopGroup = Lua.Class(ALittle.LoopObject, "ALittle.LoopGroup")

function LoopGroup:Ctor()
	___rawset(self, "_total_count", 0)
	___rawset(self, "_complete_count", 0)
	___rawset(self, "_loop_updaters", {})
	___rawset(self, "_complete_updaters", {})
	___rawset(self, "_complete_callback", nil)
end

function LoopGroup.__getter:complete_callback()
	return self._complete_callback
end

function LoopGroup.__setter:complete_callback(value)
	self._complete_callback = value
end

function LoopGroup.__getter:total_count()
	return self._total_count
end

function LoopGroup:AddUpdater(value)
	if value == nil then
		return
	end
	if self._complete_updaters[value] or self._loop_updaters[value] then
		return
	end
	self._loop_updaters[value] = true
	self._total_count = self._total_count + 1
end

function LoopGroup:RemoveUpdater(value)
	if self._loop_updaters[value] then
		self._loop_updaters[value] = nil
		self._total_count = self._total_count - 1
		return
	end
	if self._complete_updaters[value] then
		self._complete_updaters[value] = nil
		self._total_count = self._total_count - 1
		self._complete_count = self._complete_count - 1
	end
end

function LoopGroup:Reset()
	self._complete_count = 0
	for updater, v in ___pairs(self._loop_updaters) do
		updater:Reset()
	end
	for updater, v in ___pairs(self._complete_updaters) do
		self._loop_updaters[updater] = true
		updater:Reset()
	end
end

function LoopGroup:SetTime(time)
	if self._total_count == 0 then
		return time, true
	end
	local total_updater = {}
	for updater, v in ___pairs(self._loop_updaters) do
		total_updater[updater] = true
	end
	for updater, v in ___pairs(self._complete_updaters) do
		total_updater[updater] = true
	end
	self._loop_updaters = {}
	self._complete_updaters = {}
	self._complete_count = 0
	local min_time = nil
	for updater, v in ___pairs(total_updater) do
		local remain_time, completed = updater:SetTime(time)
		if completed then
			self._complete_updaters[updater] = true
			self._complete_count = self._complete_count + 1
		else
			self._loop_updaters[updater] = true
		end
		if min_time == nil or min_time > remain_time then
			min_time = remain_time
		end
	end
	return min_time, self._complete_count == self._total_count
end

function LoopGroup:IsCompleted()
	return self._complete_count >= self._total_count
end

function LoopGroup:Completed()
	if self._complete_callback ~= nil then
		self._complete_callback()
	end
end

function LoopGroup:SetCompleted()
	self._complete_count = self._total_count
	for updater, v in ___pairs(self._loop_updaters) do
		updater:SetCompleted()
		self._complete_updaters[updater] = true
	end
	self._loop_updaters = {}
end

function LoopGroup:Update(frame_time)
	if self._complete_count >= self._total_count then
		return
	end
	local remove_map = {}
	for updater, v in ___pairs(self._loop_updaters) do
		if updater:IsCompleted() then
			remove_map[updater] = true
		else
			updater:Update(frame_time)
		end
	end
	for updater, v in ___pairs(remove_map) do
		self._loop_updaters[updater] = nil
		self._complete_updaters[updater] = true
		self._complete_count = self._complete_count + 1
		updater:Completed()
	end
end

File: ALittle/Module/Std/JSScript/Std/Loop/LoopGroup.js
{
if (typeof ALittle === "undefined") window.ALittle = {};


if (ALittle.LoopObject === undefined) throw new Error(" extends class:ALittle.LoopObject is undefined");
ALittle.LoopGroup = JavaScript.Class(ALittle.LoopObject, {
	Ctor : function() {
		this._total_count = 0;
		this._complete_count = 0;
		this._loop_updaters = new Map();
		this._complete_updaters = new Map();
		this._complete_callback = undefined;
	},
	get complete_callback() {
		return this._complete_callback;
	},
	set complete_callback(value) {
		this._complete_callback = value;
	},
	get total_count() {
		return this._total_count;
	},
	AddUpdater : function(value) {
		if (value === undefined) {
			return;
		}
		if (this._complete_updaters.get(value) || this._loop_updaters.get(value)) {
			return;
		}
		this._loop_updaters.set(value, true);
		++ this._total_count;
	},
	RemoveUpdater : function(value) {
		if (this._loop_updaters.get(value)) {
			this._loop_updaters.delete(value);
			-- this._total_count;
			return;
		}
		if (this._complete_updaters.get(value)) {
			this._complete_updaters.delete(value);
			-- this._total_count;
			-- this._complete_count;
		}
	},
	Reset : function() {
		this._complete_count = 0;
		for (let [updater, v] of this._loop_updaters) {
			if (v === undefined) continue;
			updater.Reset();
		}
		for (let [updater, v] of this._complete_updaters) {
			if (v === undefined) continue;
			this._loop_updaters.set(updater, true);
			updater.Reset();
		}
	},
	SetTime : function(time) {
		if (this._total_count === 0) {
			return [time, true];
		}
		let total_updater = new Map();
		for (let [updater, v] of this._loop_updaters) {
			if (v === undefined) continue;
			total_updater.set(updater, true);
		}
		for (let [updater, v] of this._complete_updaters) {
			if (v === undefined) continue;
			total_updater.set(updater, true);
		}
		this._loop_updaters = new Map();
		this._complete_updaters = new Map();
		this._complete_count = 0;
		let min_time = undefined;
		for (let [updater, v] of total_updater) {
			if (v === undefined) continue;
			let [remain_time, completed] = updater.SetTime(time);
			if (completed) {
				this._complete_updaters.set(updater, true);
				++ this._complete_count;
			} else {
				this._loop_updaters.set(updater, true);
			}
			if (min_time === undefined || min_time > remain_time) {
				min_time = remain_time;
			}
		}
		return [min_time, this._complete_count === this._total_count];
	},
	IsCompleted : function() {
		return this._complete_count >= this._total_count;
	},
	Completed : function() {
		if (this._complete_callback !== undefined) {
			this._complete_callback();
		}
	},
	SetCompleted : function() {
		this._complete_count = this._total_count;
		for (let [updater, v] of this._loop_updaters) {
			if (v === undefined) continue;
			updater.SetCompleted();
			this._complete_updaters.set(updater, true);
		}
		this._loop_updaters = new Map();
	},
	Update : function(frame_time) {
		if (this._complete_count >= this._total_count) {
			return;
		}
		let remove_map = new Map();
		for (let [updater, v] of this._loop_updaters) {
			if (v === undefined) continue;
			if (updater.IsCompleted()) {
				remove_map.set(updater, true);
			} else {
				updater.Update(frame_time);
			}
		}
		for (let [updater, v] of remove_map) {
			if (v === undefined) continue;
			this._loop_updaters.delete(updater);
			this._complete_updaters.set(updater, true);
			++ this._complete_count;
			updater.Completed();
		}
	},
}, "ALittle.LoopGroup");

}